File: apps/backend-api/app/core/research.py
"""
Anonymous identity for the research study.

The tool is going to university journalism students through a WhatsApp group,
so nearly every session is anonymous and there is no recruitment step at which
a participant code could be issued. Runs are grouped by a device id instead: a
UUID the browser generates on first visit, keeps in localStorage, and sends as
`X-Anon-Id`, with a per-session id in `X-Session-Id`.

Why not IP, which was the original suggestion
---------------------------------------------
IP fails in both directions at once here. Sri Lankan mobile carriers run CGNAT
and a university network is a single NAT address, so one IP over-merges an
entire class into one "participant". The same student moving from campus wifi
to 4G to home broadband under-merges into three. Grouping research data that
way produces clusters that look real and are not, which is worse than having no
grouping at all. `ip_hash` therefore keeps its existing job — rate limiting —
and nothing else.

What a device id is not
-----------------------
It is not a person. It does not survive a cleared browser or a private window,
and a shared newsroom machine merges everyone who uses it. It is good enough to
group one person's work session and no stronger claim should be made of it in
the writeup.
"""

import re
from dataclasses import dataclass

from fastapi import Request

from app.schemas.auth import AuthUser

# Opaque client-generated ids. Length-capped and character-restricted because
# these reach the database: a header is attacker-controlled, and "it is only
# an analytics field" is how junk ends up indexed. Anything not matching is
# dropped rather than rejected — a malformed header must never fail a
# journalist's grammar check.
_ID = re.compile(r"[A-Za-z0-9._-]{8,64}")

ANON_HEADER = "X-Anon-Id"
SESSION_HEADER = "X-Session-Id"


@dataclass(frozen=True)
class Actor:
    """Who made a request, for research grouping."""

    user_id: str | None
    anon_id: str | None
    session_id: str | None

    @property
    def is_known(self) -> bool:
        """True when the run can be attributed to somebody at all."""
        return bool(self.user_id or self.anon_id)

    def stamp(self) -> dict[str, str | None]:
        """The identity columns, ready to merge into a row."""
        return {
            "user_id": self.user_id,
            "anon_id": self.anon_id,
            "session_id": self.session_id,
        }


def _clean(raw: str | None) -> str | None:
    if not raw:
        return None
    raw = raw.strip()
    return raw if _ID.fullmatch(raw) else None


def actor_from(request: Request, user: AuthUser | None) -> Actor:
    """
    Resolve the caller.

    A signed-in user still carries their device id, so a student who signs in
    part-way through does not appear as two unrelated participants.
    """
    return Actor(
        user_id=user.id if user else None,
        anon_id=_clean(request.headers.get(ANON_HEADER)),
        session_id=_clean(request.headers.get(SESSION_HEADER)),
    )

File: apps/web-app/src/lib/research.js
/**
 * Anonymous identity for the research study.
 *
 * The tool reaches university journalism students through a WhatsApp group, so
 * there is no recruitment step at which a participant code could be handed out
 * and almost every session is anonymous. Runs are grouped by a device id
 * generated here on first visit and kept in localStorage.
 *
 * Deliberately not the IP address, which the backend sees anyway: Sri Lankan
 * mobile carriers use CGNAT and a campus network is one NAT address, so an IP
 * merges a whole class into one "participant" while splitting one student
 * across wifi, 4G and home broadband. A device id is wrong in neither
 * direction.
 *
 * What it is not: a person. It does not survive a cleared browser or a private
 * window, and a shared machine merges everyone who uses it.
 */

const DEVICE_KEY = 'sinai.device_id';
const SESSION_KEY = 'sinai.session_id';

// A session is one sitting. Half an hour of inactivity ends it, so the study
// can talk about "a work session" rather than one undivided blob per device.
const SESSION_IDLE_MS = 30 * 60 * 1000;

function newId(prefix) {
  const uuid =
    globalThis.crypto?.randomUUID?.() ??
    // Older Safari on the phones a lot of students will be using has no
    // randomUUID. Collision risk across a study cohort is negligible and a
    // duplicate id costs a merged pair of sessions, not a broken request.
    `${Date.now().toString(36)}${Math.random().toString(36).slice(2, 12)}`;
  return `${prefix}-${uuid.replace(/-/g, '')}`;
}

/**
 * Read a key, tolerating storage being unavailable.
 *
 * Private browsing and blocked third-party storage both make localStorage
 * throw on access rather than return null. Losing the id costs the study one
 * ungrouped session; letting the exception escape would break the app, so
 * every path here degrades to "no id" instead.
 */
function read(key) {
  try {
    return window.localStorage.getItem(key);
  } catch {
    return null;
  }
}

function write(key, value) {
  try {
    window.localStorage.setItem(key, value);
  } catch {
    /* storage unavailable — this request simply goes ungrouped */
  }
}

export function getDeviceId() {
  let id = read(DEVICE_KEY);
  if (!id) {
    id = newId('d');
    write(DEVICE_KEY, id);
  }
  return id;
}

/** Current session id, rotating after SESSION_IDLE_MS of inactivity. */
export function getSessionId() {
  const now = Date.now();
  let id = null;
  let last = 0;
  try {
    const raw = read(SESSION_KEY);
    if (raw) ({ id, last } = JSON.parse(raw));
  } catch {
    id = null;
  }
  if (!id || now - last > SESSION_IDLE_MS) id = newId('s');
  write(SESSION_KEY, JSON.stringify({ id, last: now }));
  return id;
}

/** Headers identifying this device and sitting, for every API call. */
export function researchHeaders() {
  try {
    return { 'X-Anon-Id': getDeviceId(), 'X-Session-Id': getSessionId() };
  } catch {
    return {};
  }
}
